results.tex

```diff
@@ -39,20 +39,20 @@ \subsubsection{Partitioning and distribution algorithms}
 
 \subsection{Currently used implementations}
 \subsubsection{Generic distributed system frameworks}
-\subsubsection{Domain-specific implementations}
-\subsubsection{Current challenges}
-
-
-
+\input{./sections/domain_specific_implementations.tex}
 
+\subsubsection{Current challenges}
 
+\paragraph{Performance}
 
+A tradeoff that’s seen frequently is the reduction of wall-clock time at cost of significantly increased total processing time. When compute resources are affordable enough, many real-world use cases of machine learning benefit most from being trained rapidly. The fact that this often implies a large multiple in total compute resources (and the associated energy consumption) is not as important, given the revenues and cost-reductions that applying machine learning can effect.  A good example of this is the DistBelief paper, where wall clock time speedup factors are achieved by increasing the number of machines quadratically or worse, and data parallelism only comes out ahead of GPUs when using more than 20 machines. Distributed use of GPUs, as in Tensorflow, has better properties, 
```

Comment:
Maybe replace `and data parallelism only comes out ahead of GPUs when using more than 20 machines` by `and the performance achieved only surpasses the performance of GPUs when more than 20 machines are used`